bcda/auth/alpha.go

```diff
@@ -140,39 +140,52 @@ func (p AlphaAuthPlugin) RevokeClientCredentials(params []byte) error {
 	return nil
 }
 
-// generate a token for the id (which user? just have a single "user" (alpha2, alpha3, ...) per test cycle?)
-// params are currently acoId and ttl; not going to introduce user until we have clear use cases
+// generate a token for the ACO, either for a specified UserID or (if not provided) any user in the ACO
 func (p AlphaAuthPlugin) RequestAccessToken(creds Credentials, ttl int) (Token, error) {
-	db := database.GetGORMDbConnection()
-	defer database.Close(db)
-
+	var userUUID, acoUUID uuid.UUID
+	var user models.User
+	var err error
 	token := Token{}
 
-	acoUUID := creds.ClientID
-	if acoUUID == "" {
-		return token, errors.New("no ACO ID provided")
+	if creds.UserID == "" && creds.ClientID == "" {
+		return token, fmt.Errorf("must provide either UserID or ClientID")
 	}
 
-	aco, err := getACOFromDB(acoUUID)
-	if err != nil {
-		return token, err
+	if ttl < 0 {
+		return token, fmt.Errorf("invalid TTL: %d", ttl)
 	}
 
-	// I arbitrarily decided to use the first user. An alternative would be to make a specific user
-	// that represents the client. I have no strong opinion here other than not creating stuff in the db
-	// unless we're willing to live with it forever.
-	var user models.User
-	if err = db.First(&user, "aco_id = ?", aco.UUID).Error; err != nil {
-		return token, errors.New("no user found for " + aco.UUID.String())
-	}
+	db := database.GetGORMDbConnection()
+	defer database.Close(db)
 
-	if ttl < 0 {
-		return token, fmt.Errorf("invalid TTL: %d", ttl)
+	// Get both acoUUID and userUUID based on creds.UserID if provided
+	if creds.UserID != "" {
+		if db.First(&user, "UUID = ?", creds.UserID).RecordNotFound() {
+			return token, fmt.Errorf("unable to locate User with id of %s", creds.UserID)
```

Comment:
you can't return nil as type `Token` from this func though since the func signature declares a `Token` as the first output.  so it's kind of interesting that you can return nil instead of type `Error`, which is definitely that idiomatic Go we love.  that is allowed because the zero value of the `error` type is actually `nil`.  `Token`'s nil value is `Token{}` for contrast.  kinda cool!

Reply:
and looks like the zero value of structs is just the struct with zero values for each field: https://play.golang.org/p/QDohNnjEpp0